### PR TITLE
Lowering Gemm

### PR DESCRIPTION
**Pull request by tungld**
This patch is to lower GEMM operator.

### CODE REVIEW

File: src/pass/lower_frontend_to_krnl.cpp [hunk 1]
```diff
@@ -1100,6 +1100,219 @@ struct ONNXReshapeOpLowering : public ConversionPattern {
...
+    rewriter.create<StoreOp>(loc, Y, alloc, loopMNIVs);
+
+    // Insert instructions to do matrix multiplication: A*B
+    Block &matmulIterationBlock = matmulIterateOp.bodyRegion().front();
+    rewriter.setInsertionPointToStart(&matmulIterationBlock);
+
+    // Induction variables
+    SmallVector<Value, 4> loopKIVs, loopAIVs, loopBIVs;
+    for (auto arg : matmulIterationBlock.getArguments())
+      loopKIVs.push_back(arg);
```
Comment (doru1004) [inline src/pass/lower_frontend_to_krnl.cpp]: emplace_back

File: src/pass/lower_frontend_to_krnl.cpp [hunk 2]
```diff
@@ -1100,6 +1100,219 @@ struct ONNXReshapeOpLowering : public ConversionPattern {
...
+    rewriter.setInsertionPoint(optimizedLoopsOp);
+
+    // Insert instructions inside the outer loop.
+    Block &outerIterationBlock = outerIterateOp.bodyRegion().front();
+    rewriter.setInsertionPointToStart(&outerIterationBlock);
+
+    // Induction variables
+    SmallVector<Value, 4> loopMNIVs;
+    for (auto arg : outerIterationBlock.getArguments()) {
+      loopMNIVs.push_back(arg);
```
Comment (doru1004) [inline src/pass/lower_frontend_to_krnl.cpp]: emplace_back

File: src/dialect/onnx/onnx_ops.cpp [hunk 3]
```diff
@@ -350,9 +350,23 @@ void ONNXGemmOp::inferShapes() {
...
+
+  int64_t M, N, K_A, K_B;
+  M = (transA == 0) ? lhsTy.getShape()[0] : lhsTy.getShape()[1];
+  K_A = (transA == 0) ? lhsTy.getShape()[1] : lhsTy.getShape()[0];
+  N = (transB == 0) ? rhsTy.getShape()[1] : rhsTy.getShape()[0];
+  K_B = (transB == 0) ? rhsTy.getShape()[0] : rhsTy.getShape()[1];
+
+  if ((K_A != -1) and (K_B != -1) and (K_A != K_B)) {
+    emitError("Tensor shapes mismatched.");
+    return;
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/dialect/onnx/onnx_ops.cpp]: If so, the remaining statements will be executed normally and the result type will be set. Is it  redundant?

File: src/pass/lower_frontend_to_krnl.cpp [hunk 4]
```diff
@@ -1100,6 +1100,219 @@ struct ONNXReshapeOpLowering : public ConversionPattern {
...
+    Value alloc;
+    bool insertDealloc = checkInsertDealloc(op);
+    if (hasAllConstantDimensions(memRefType))
+      alloc = insertAllocAndDealloc(memRefType, loc, rewriter, insertDealloc);
+    else {
+      auto memRefShape = memRefType.getShape();
+      SmallVector<Value, 2> allocOperands;
+      if (memRefShape[0] < 0) {
+        auto dim = rewriter.create<DimOp>(loc, A, (isTransA) ? 1 : 0);
+        allocOperands.push_back(dim);
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/pass/lower_frontend_to_krnl.cpp]: Done.

File: src/dialect/onnx/onnx_ops.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (tjingrant) [reply src/dialect/onnx/onnx_ops.cpp]: I agree with @doru1004 's comment. Exception throwing automatically terminates program execution. Assuming emitError will throw an exception, the following statement will not be executed.

[the remark this answers is not in the document]
Comment (tungld) [reply src/dialect/onnx/onnx_ops.cpp]: Not sure why but in my environment, the program did not terminate if I removed the `return`, In such a case, I could even emit another error with another `emitError` statement.

[the remark this answers is not in the document]
Comment (tungld) [reply src/dialect/onnx/onnx_ops.cpp]: Example:
Input program with mismatched type:
```
func @add(%arg0 : tensor<6x10xf32>, %arg1 : tensor<5x10xf32>, %arg2: tensor<f32>) -> tensor<*xf32> {
  %0 ="onnx.Gemm"(%arg0, %arg1, %arg2) {alpha = 1.0 : f32, beta = 1.0 : f32, transA = 1, transB = 0} : (tensor<6x10xf32>, tensor<5x10xf32>, tensor<f32>) -> tensor<*xf32    >
"std.return"(%0) : (tensor<*xf32>) -> ()
 }
```

If I removed the `return`, I still got this result:
(Command: `onnf-opt --shape-inference --lower-frontend`)
```
(python3.6) tungl@s7fp1:~/dl/ONNF/build$ onnf-opt --shape-inference ~/dl/dlc-misc/model.mlir --lower-frontend
/home/tungld/dl/dlc-misc/model.mlir:50:7: error: Tensor shapes mismatched.
  %0 ="onnx.Gemm"(%arg0, %arg1, %arg2) {alpha = 1.0 : f32, beta = 1.0 : f32, transA = 1, transB = 0} : (tensor<6x10xf32>, tensor<5x10xf32>, tensor<f32>) -> tensor<*xf32>
      ^
#map0 = affine_map<() -> (0)>
#map1 = affine_map<() -> (6)>
#map2 = affine_map<() -> (10)>


module {
  func @add(%arg0: memref<6x10xf32>, %arg1: memref<5x10xf32>, %arg2: memref<f32>) -> memref<10x10xf32> {
    %0 = alloc() : memref<10x10xf32>
    %cst = constant 1.000000e+00 : f32
    %cst_0 = constant 1.000000e+00 : f32
    %1:3 = krnl.define_loops 3
    %2:3 = krnl.optimize_loops  {
      krnl.return_loops %1#0, %1#1, %1#2
    } : () -> (!krnl.loop, !krnl.loop, !krnl.loop)
    krnl.iterate(%2#0, %2#1) with (%1#0 -> %arg3 = 0 to 10, %1#1 -> %arg4 = 0 to 10) {
      %cst_1 = constant 0.000000e+00 : f32
      store %cst_1, %0[%arg3, %arg4] : memref<10x10xf32>
      krnl.iterate(%2#2) with (%1#2 -> %arg5 = 0 to 6) {
        %8 = load %arg0[%arg5, %arg3] : memref<6x10xf32>
        %9 = load %arg1[%arg5, %arg4] : memref<5x10xf32>
        %10 = load %0[%arg3, %arg4] : memref<10x10xf32>
        %11 = mulf %8, %9 : f32
        %12 = addf %10, %11 : f32
        store %12, %0[%arg3, %arg4] : memref<10x10xf32>
      }
      %3 = load %arg2[] : memref<f32>
      %4 = load %0[%arg3, %arg4] : memref<10x10xf32>
      %5 = mulf %cst, %4 : f32
      %6 = mulf %cst_0, %3 : f32
      %7 = addf %5, %6 : f32
      store %7, %0[%arg3, %arg4] : memref<10x10xf32>
    }
    return %0 : memref<10x10xf32>
  }
}
```

[the remark this answers is not in the document]
Comment (tungld) [reply src/dialect/onnx/onnx_ops.cpp]: If I emited two errors like this:
```
if ((K_A != -1) and (K_B != -1) and (K_A != K_B)) {
  emitError("Tensor shapes mismatched.");
}

emitError("Tensor shapes mismatched.");
```
I saw two error messages, and the program did not terminate.
```
/home/tungld/dl/dlc-misc/model.mlir:50:7: error: Tensor shapes mismatched.
  %0 ="onnx.Gemm"(%arg0, %arg1, %arg2) {alpha = 1.0 : f32, beta = 1.0 : f32, transA = 1, transB = 0} : (tensor<6x10xf32>, tensor<5x10xf32>, tensor<f32>) -> tensor<*xf32>
      ^
/home/tungld/dl/dlc-misc/model.mlir:50:7: error: Tensor shapes mismatched.
  %0 ="onnx.Gemm"(%arg0, %arg1, %arg2) {alpha = 1.0 : f32, beta = 1.0 : f32, transA = 1, transB = 0} : (tensor<6x10xf32>, tensor<5x10xf32>, tensor<f32>) -> tensor<*xf32>
      ^
#map0 = affine_map<() -> (0)>
#map1 = affine_map<() -> (6)>
#map2 = affine_map<() -> (10)>
...
```

[the remark this answers is not in the document]
Comment (tungld) [reply src/dialect/onnx/onnx_ops.cpp]: I am not sure it was because of my environment or MLIR behavior.

**Pull request merged**